xarray/core/concat.py

```diff
@@ -506,22 +512,52 @@ def ensure_common_dims(vars):
                 var = var.set_dims(common_dims, common_shape)
             yield var
 
-    # stack up each variable to fill-out the dataset (in order)
+    # stack up each variable and/or index to fill-out the dataset (in order)
     # n.b. this loop preserves variable order, needed for groupby.
-    for k in datasets[0].variables:
-        if k in concat_over:
+    for name in datasets[0].variables:
+        if name in concat_over and name not in result_indexes:
             try:
-                vars = ensure_common_dims([ds[k].variable for ds in datasets])
+                vars = ensure_common_dims([ds[name].variable for ds in datasets])
             except KeyError:
-                raise ValueError(f"{k!r} is not present in all datasets.")
-            combined = concat_vars(vars, dim, positions, combine_attrs=combine_attrs)
-            assert isinstance(combined, Variable)
-            result_vars[k] = combined
-        elif k in result_vars:
+                raise ValueError(f"{name!r} is not present in all datasets.")
+
+            # Try concatenate the indexes first, silently fallback to concatenate
+            # the variables when no index is found on all datasets or when the
+            # 1st index doesn't implement concat.
```

Comment:
Could we eliminate the fallback case here?

I guess `concat` is internally pretty robust at present, but requiring matching indexes would keep things simpler.

Reply:
I agree that it would keep things simpler, at least for now.

More generally, I'm wondering if requiring matching indexes for alignment, merge and concat wouldn't be too restrictive / not practical.

For example, let's say that we want to concatenate a sequence of Xarray objects along a dimension `x` where some of them have a geospatial raster (meta-)index, e.g., for dimensions `x` and `y`, and all the others have decoupled pandas indexes for `x` and `y`. Which behavior do we want?

- A. Raise an error (no matching index along `x`, which is actually raised when aligning the objects before concatenation)
- B. Concatenate the coordinate variables along `x`, merge along the other dimensions and return the concatenated Dataset or DataArray with default pandas indexes

Option B would perhaps be more convenient, but significantly harder to implement! Although it is also easy to write `xr.concat([ds.reset_index("x") for ds in datasets], "x")`.

Back to your comment, I guess that you are referring to the 1st fallback and not the one when `index.concat` is not implemented?

Reply:
I think it would probably be OK to remove both fall-backs here:

1. Require that all indexes exist if at least one exists.
2. Require implementing `Index.concat`.

Right now you are simply dropping indexes in this case, which seems non-ideal.

Reply:
Removing both fall-backs breaks this test:

```python
def test_concat_promote_shape(self) -> None:
    # mixed dims within variables
    objs = [Dataset({}, {"x": 0}), Dataset({"x": [1]})]
    actual = concat(objs, "x")
```

What should we do?

- A. Implicitly create a ``PandasIndex`` for any scalar coordinate to concatenate, then concatenate the indexes with `Index.concat`
- B. Fallback to `Variable.concat` (it will create a new, default `PandasIndex` from the concatenated variable, when creating the resulting Dataset object)

Option A will likely not work if the other existing indexes are not all `PandasIndex` instances. Option B is a bit more flexible, although we should be careful when to fall-back as I agree dropping indexes is non-ideal, e.g., the example below:

```python
objs = [xr.Dataset(coords={"z": ("y", [2, 3])}), xr.Dataset(coords={"z": ("x", [4, 5])})]
# Let's set an index for the 1st object only (no public API yet)
objs[0]._indexes["z"] = PandasIndex([2, 3], "y")

# this should probably raise an error
# instead of concatenate "z", promote its dims to ("y", "x"), and drop the index
actual = xr.concat(objs, "x")
```

Reply:
OK, I'm pretty sure I wrote that test! And now I'm doubting whether that behavior is really desirable, though I'm sure it may be more convenient than getting an error.

I would consider limiting the behavior to (A) for now (only supporting PandasIndex), and maybe deprecating it later?